### PR TITLE
Enhancement/improve gpu selection

### PR DESCRIPTION
**Pull request by HenriChabert**

### CODE REVIEW

File: js/extractRecipeController.js [hunk 1]
```diff
@@ -48,15 +49,15 @@ app.controller('extractRecipeController', function($scope) {
         });
     };
     
-    var handleGPU = function(data) {
+    let handleGPU = function(data) {
         $scope.gpuList = data["gpu_list"];
         $scope.canUseGPU = data["can_use_gpu"];
         $scope.gpuUsage = data["gpu_usage_choices"];
-        $scope.gpuMemory = data["gpu_memory_choices"];
+        $scope.gpuMemoryAllocationMode = data["gpu_memory_allocation_mode"];
```
Comment (alexcombessie) [inline js/extractRecipeController.js:56]: Are you sure that this will not break existing recipes? Since there are no migration steps, changing a parameter is a risky operation.

File: python-lib/dku_deeplearning_image/constants.py [hunk 2]
```diff
@@ -28,3 +28,6 @@
 DEFAULT_PRED_MIN_THRESHOLD = 0
 COMPILE_LOSS_FUNCTION = 'categorical_crossentropy'
 COMPILE_METRICS = ["accuracy"]
+
+GPU_MEMORY_LIMIT = "memory_limit"
```
Comment (alexcombessie) [inline python-lib/dku_deeplearning_image/constants.py:32]: I would advise doing an Enum or class constants to denote that these are two possible modes of the same parameter

File: python-lib/dku_deeplearning_image/utils.py [hunk 3]
```diff
@@ -77,25 +78,39 @@ def can_use_gpu():
     return len(tf.config.experimental.list_physical_devices('GPU')) > 0
 
 
-def set_gpu_options(should_use_gpu, gpu_list, memory_limit):
+def set_gpu_options(should_use_gpu, gpu_list, gpu_memory_allocation_mode, memory_limit_ratio=None):
```
Comment (alexcombessie) [inline python-lib/dku_deeplearning_image/utils.py:81]: Just a nice to have, but type hints would help understand what you expect here

File: python-lib/dku_deeplearning_image/utils.py [hunk 4]
```diff
@@ -77,25 +78,39 @@ def can_use_gpu():
...
     log_info("load_gpu_options")
```
Comment (alexcombessie) [inline python-lib/dku_deeplearning_image/utils.py:82]: Typo or leftover from a debug session? Log messages should be a bit more readable

File: python-lib/dku_deeplearning_image/utils.py [hunk 5]
```diff
@@ -77,25 +78,39 @@ def can_use_gpu():
...
     if should_use_gpu and can_use_gpu():
         log_info("should use GPU")
         gpus = tf.config.experimental.list_physical_devices('GPU')
```
Comment (alexcombessie) [inline python-lib/dku_deeplearning_image/utils.py:85]: Perhaps better to use the official (non-experimental) API? https://www.tensorflow.org/api_docs/python/tf/config/list_physical_devices

File: python-lib/dku_deeplearning_image/utils.py [hunk 6]
```diff
@@ -77,25 +78,39 @@ def can_use_gpu():
...
+        for i, g in enumerate(gpus):
```
Comment (alexcombessie) [inline python-lib/dku_deeplearning_image/utils.py:86]: Why is this needed?

File: python-lib/dku_deeplearning_image/utils.py [hunk 7]
```diff
@@ -77,25 +78,39 @@ def can_use_gpu():
...
+            g.id = i
         gpus_to_use = [gpus[int(i)] for i in gpu_list] or gpus
-        if memory_limit:
+        log_info("GPUs on the machine: {}".format(str([g.id for g in GPUtil.getGPUs()])))
```
Comment (alexcombessie) [inline python-lib/dku_deeplearning_image/utils.py:89]: You don't need to `str`, format takes care of that for you. Bonus: you could use f"" string formatting

File: python-lib/dku_deeplearning_image/utils.py [hunk 8]
```diff
@@ -77,25 +78,39 @@ def can_use_gpu():
...
+        log_info("Will use the following GPUs: {}".format(gpus_to_use))
+        if gpu_memory_allocation_mode == constants.GPU_MEMORY_LIMIT and memory_limit_ratio:
             for gpu in gpus_to_use:
+                memory_limit = calculate_gpu_memory_allocation(memory_limit_ratio, gpu)
+                log_info("Restraining GPU(s) {} to {} Mo".format(gpu, str(memory_limit)))
```
Comment (alexcombessie) [inline python-lib/dku_deeplearning_image/utils.py:94]: What if GPUs have different memory sizes? Perhaps better to log only the ratio?

Comment (alexcombessie) [reply python-lib/dku_deeplearning_image/utils.py:94]: Also, no need to `str` and proposition to use f"" string formatting

File: python-lib/dku_deeplearning_image/utils.py [hunk 9]
```diff
@@ -77,25 +78,39 @@ def can_use_gpu():
...
                 tf.config.experimental.set_virtual_device_configuration(
                     gpu,
                     [tf.config.experimental.VirtualDeviceConfiguration(memory_limit=int(memory_limit))]
```
Comment (alexcombessie) [inline python-lib/dku_deeplearning_image/utils.py:97]: I would advise against using anything from tf.config.experimental, but only tf.config.

File: python-lib/dku_deeplearning_image/utils.py [hunk 10]
```diff
@@ -10,6 +10,7 @@
 from collections import OrderedDict
 import numpy as np
 from datetime import datetime
+import GPUtil
```
Comment (alexcombessie) [inline python-lib/dku_deeplearning_image/utils.py:13]: Code structure suggestion, why not put all the GPU-related functions from utils.py in GPUtil?

File: resource/helpers/recipes-helper.py [hunk 11]
```diff
@@ -104,13 +104,17 @@ def add_gpu_options_to_resp(response):
             'value': 'custom'
         }
     ]
-    response["gpu_memory_choices"] = [
+    response["gpu_memory_allocation_mode"] = [
         {
             'label': 'No limitation',
             'value': 'all'
```
Comment (alexcombessie) [inline resource/helpers/recipes-helper.py:110]: +1 for doing an enum for possible values

File: resource/helpers/recipes-helper.py [hunk 12]
```diff
@@ -104,13 +104,17 @@ def add_gpu_options_to_resp(response):
...
         },
+        {
+            'label': 'Allocate only as much GPU memory as needed',
```
Comment (alexcombessie) [inline resource/helpers/recipes-helper.py:113]: I would advise a shorter label, to make it fit in the SELECT box. How about "Allocate memory automatically"?

File: resource/templates/extract-recipe-template.html [hunk 13]
```diff
@@ -48,18 +48,18 @@ <h2 class="settings-section-title">GPU</h2>
                <label class="control-label">GPU Memory</label>
                <div class="controls">
                   <select dku-bs-select
-                     name="gpu_memory"
-                     ng-model="config.gpu_memory"
-                     ng-options="gm.value as gm.label for gm in gpuMemory">
+                     name="gpu_memory_allocation_mode"
```
Comment (alexcombessie) [inline resource/templates/extract-recipe-template.html:51]: There is some duplication of code here (same as for JS files). Could you have a dedicated template for GPU-related stuff and call that template from each recipe template?

File: resource/templates/extract-recipe-template.html [hunk 14]
```diff
@@ -48,18 +48,18 @@ <h2 class="settings-section-title">GPU</h2>
...
+                     ng-model="config.gpu_memory_allocation_mode"
+                     ng-options="gm.value as gm.label for gm in gpuMemoryAllocationMode">
                   </select>
                </div>
             </div>
-            <div ng-if="config.gpu_memory == 'custom'">
+            <div ng-if="config.gpu_memory_allocation_mode == 'memory_limit'">
                <div class="control-group">
-                  <label class="control-label">Memory limit per GPU</label>
+                  <label class="control-label">Memory allocation</label>
```
Comment (alexcombessie) [inline resource/templates/extract-recipe-template.html:59]: Suggestion: for the user to better understand that this parameter is linked to the "GPU Memory" settings, you could use the same arrow symbol we used here: https://github.com/dataiku/dss-plugin-timeseries-forecast/blob/04f4dd661370695aac4e97f22f938abf713ab65c/custom-recipes/timeseries-forecast-1-train-evaluate/recipe.json#L466
```suggestion
                  <label class="control-label"> ↳ Memory allocation</label>
```

Comment (alexcombessie) [reply resource/templates/extract-recipe-template.html:59]: Also, not sure the label "Memory allocation" is easier to understand for the user than the previous term "Memory limit per GPU". In particular, the select options above refers to a "custom memory limit".

**Review by alexcombessie — CHANGES_REQUESTED** [review]
Hi @HenriChabert,

Good work! I have left a few nitpick suggestions. The big change I am suggesting is to avoid code duplication across recipes. Since all the GPU settings are common to the 3 recipes, you would be better off centralizing them in a single JS file and a single HTML template, called by the recipes.

Otherwise, a few UX comments not related directly to this PR:
- (Nitpick #PTSD) The parameter names don't follow a consistent case.
  - Most of the time it's "This case": "Label column", " Layer(s) to retrain", "Train ratio".
  - But sometimes it's "Random Seed", "Image Shape"
  - I would recommend to normalize it to "This case", as it's the majority case used in the rest of DSS.
- (Nice to have) Could the plugin be categorized as "Computer Vision" to make it easier to find? Customers now have a lot of plugins, it helps them to have categories.
- (Nice to have) Why not have the plugin selectable from the flow?
- (To be discussed) The recipe descriptions are very long, could they be shortened to 1-2 lines? For instance, input/output descriptions or technical usage recommendations would better be in the input/output tooltips and in the documentation.
- (To be discussed) Should the (v2) be mentioned everywhere? An alternative could be to mark the previous version as legacy, and present the new version as the current one.
- (Genuine question) How will you handle the GPU / non-GPU requirements with a single plugin id?

**Comment by alexcombessie** [comment]
Other weird behavior I am getting on dku11. With the following settings:
Screenshot 2021-03-16 at 12 12 23
I am getting this usage:
Screenshot 2021-03-16 at 12 12 51
Source: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/jobs/Build_retrained_with_as_much_gpu_as_needed_2021-03-16T11-11-19.276/?hideFlow=true

Now if I relaunch the recipe with these settings:
Screenshot 2021-03-16 at 12 14 18
I get this usage:
Screenshot 2021-03-16 at 12 14 12
Source: http://localhost:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/recipes/compute_jb87eTC0_1_1/

Isn't there something wrong? The ratio chosen in the recipe is not reflected in the actual usage from nvidia-smi

**Comment by alexcombessie** [comment]
... and I found a UX 🐛 !

https://user-images.githubusercontent.com/4739848/111301271-4d76be80-8652-11eb-85fe-b45a8f7a7907.mov

The GPU button cannot be deselected

File: js/extractRecipeController.js [hunk 15]
```diff
@@ -48,15 +49,15 @@ app.controller('extractRecipeController', function($scope) {
         });
     };
     
-    var handleGPU = function(data) {
+    let handleGPU = function(data) {
```
[the remark this answers is not in the document]
Comment (HenriChabert) [reply js/extractRecipeController.js]: Well, you pinpointed what I have postponed since the beginning ;) But you are right, I refactored all that in c06d932

File: python-lib/dku_deeplearning_image/constants.py (hunk 2 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/constants.py:32]: Good idea, will do that in another PR to refactor all the constants needing an Enum

File: python-lib/dku_deeplearning_image/utils.py (hunk 3 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:81]: True, will also dedicate a PR to typing and documentation

File: python-lib/dku_deeplearning_image/utils.py (hunk 4 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: Fixed in a68a2cf

File: python-lib/dku_deeplearning_image/utils.py (hunk 5 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: True, however I use tf 2.2 here so, although this function is available in non experimental modes, it is not the case on all functions. Fixed in 36eb1ba

File: python-lib/dku_deeplearning_image/utils.py (hunk 6 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:86]: It is useful to map the coming from tf with function `tf.config.list_physical_devices('GPU')` to gpus from GPUtil. I then use these Ids in `calculate_gpu_memory_allocation(memory_limit_ratio, gpu_to_use)`

File: python-lib/dku_deeplearning_image/utils.py (hunk 7 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: I am always a little bit scared of using f string syntax, but it works for all Python 3 versions right?

File: python-lib/dku_deeplearning_image/utils.py (hunk 8 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: Prefere to do both to see how conversions worked. As I display this for each GPU, there are no problem if each GPU has a different memory size.

File: python-lib/dku_deeplearning_image/utils.py (hunk 9 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:97]: Mmmh this function typically is not available in the not experimental library. However, it is used here https://www.tensorflow.org/guide/gpu
I would rather not go on tf 2.4 because of Cuda versions

File: python-lib/dku_deeplearning_image/utils.py (hunk 10 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:13]: Actually, GPUUtil is an external library

File: resource/helpers/recipes-helper.py (hunk 11 above)
Comment (HenriChabert) [reply resource/helpers/recipes-helper.py:118]: Yop, will do in a future PR

File: resource/helpers/recipes-helper.py (hunk 12 above)
Comment (HenriChabert) [reply resource/helpers/recipes-helper.py]: You are right, fixed in 2f16f21

File: resource/templates/extract-recipe-template.html (hunk 13 above)
Comment (HenriChabert) [reply resource/templates/extract-recipe-template.html]: Fixed in c06d932

File: resource/templates/extract-recipe-template.html (hunk 14 above)
Comment (HenriChabert) [reply resource/templates/extract-recipe-template.html]: True and true. Fixed in 419bc6e

**Comment by HenriChabert** [comment]
> ... and I found a UX 🐛 !
> 
>  Screen.Recording.2021-03-16.at.12.21.52.mov 
> The GPU button cannot be deselected
> 
> Source: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/recipes/compute_jb87eTC0_1_1/

About this, I can't replicate so I assume the refacto has fixed the issue by itself 💪

**Comment by HenriChabert** [comment]
> Other weird behavior I am getting on dku11. With the following settings:
> Screenshot 2021-03-16 at 12 12 23
> I am getting this usage:
> Screenshot 2021-03-16 at 12 12 51
> Source: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/jobs/Build_retrained_with_as_much_gpu_as_needed_2021-03-16T11-11-19.276/?hideFlow=true
> 
> Now if I relaunch the recipe with these settings:
> Screenshot 2021-03-16 at 12 14 18
> I get this usage:
> Screenshot 2021-03-16 at 12 14 12
> Source: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/recipes/compute_jb87eTC0_1_1/
> 
> Isn't there something wrong? The ratio chosen in the recipe is not reflected in the actual usage from nvidia-smi

Mmmh, that is definitely right but I don't know where the problem comes from. Indeed, when looking at the logs coming from tensorflow, I get the following:
image

This means that the % has been reduced successfully but for some reason, it is a little bit higher in real. Will investigate tomorrow and ask to some people what they think.

File: js/extractRecipeController.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (alexcombessie) [reply js/extractRecipeController.js]: Much better 😄

File: python-lib/dku_deeplearning_image/utils.py (hunk 5 above)
Comment (alexcombessie) [reply python-lib/dku_deeplearning_image/utils.py]: Why not switch to tf 2.4?

File: python-lib/dku_deeplearning_image/utils.py (hunk 10 above)
Comment (alexcombessie) [reply python-lib/dku_deeplearning_image/utils.py:13]: Ah ok. Maybe just rename your module to gpu_utils.py? Assuming it contains only gpu related stuff

File: python-lib/dku_deeplearning_image/dku_constants.py [hunk 16]
```diff
@@ -37,6 +37,28 @@
 COMPILE_LOSS_FUNCTION = 'categorical_crossentropy'
 COMPILE_METRICS = ["accuracy"]
 
+# Options for custom form
+GPU_MEMORY_LIMIT = "memory_limit"
+GPU_MEMORY_GROWTH = "memory_growth"
+
+POOLING_OPTIONS = [
```
Comment (alexcombessie) [inline python-lib/dku_deeplearning_image/dku_constants.py:44]: Use enums, that's the right data structure for what you want to do

File: python-lib/dku_deeplearning_image/utils.py (hunk 6 above)
Comment (alexcombessie) [reply python-lib/dku_deeplearning_image/utils.py:86]: oki

File: python-lib/dku_deeplearning_image/utils.py (hunk 9 above)
Comment (alexcombessie) [reply python-lib/dku_deeplearning_image/utils.py:97]: Humm there are several CVE in tensorflow 2.2 which were fixed in later versions: https://github.com/tensorflow/tensorflow/releases

To me, that justifies the need to upgrade to 2.4, at the expense of older CUDA versions

**Comment by alexcombessie** [comment]
> 
> 
> > ... and I found a UX 🐛 !
> > 
> >  Screen.Recording.2021-03-16.at.12.21.52.mov 
> > The GPU button cannot be deselected
> > 
> > Source: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/recipes/compute_jb87eTC0_1_1/
> 
> About this, I can't replicate so I assume the refacto has fixed the issue by itself 💪

Hum let me test again before claiming the 🐻

File: python-lib/dku_deeplearning_image/utils.py (hunk 9 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py]: Will dedicate a new PR for that

File: python-lib/dku_deeplearning_image/utils.py (hunk 10 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/utils.py:15]: It does not, it has all utils functions for my Plugin. Do you think I should separate it?

File: python-lib/dku_deeplearning_image/dku_constants.py (hunk 16 above)
Comment (HenriChabert) [reply python-lib/dku_deeplearning_image/dku_constants.py:44]: Will do in a new PR

**Comment by HenriChabert** [comment]
> Other weird behavior I am getting on dku11. With the following settings:
> Screenshot 2021-03-16 at 12 12 23
> I am getting this usage:
> Screenshot 2021-03-16 at 12 12 51
> Source: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/jobs/Build_retrained_with_as_much_gpu_as_needed_2021-03-16T11-11-19.276/?hideFlow=true
> 
> Now if I relaunch the recipe with these settings:
> Screenshot 2021-03-16 at 12 14 18
> I get this usage:
> Screenshot 2021-03-16 at 12 14 12
> Source: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/recipes/compute_jb87eTC0_1_1/
> 
> Isn't there something wrong? The ratio chosen in the recipe is not reflected in the actual usage from nvidia-smi

So I spoke with coni and he has the same opinion than me: It must be pretty tough to really evaluate exactly the gpu for tensorflow so it makes sense that there is a gap between what we asked for and what really happens.

File: python-lib/dku_deeplearning_image/utils.py (hunk 10 above)
Comment (alexcombessie) [reply python-lib/dku_deeplearning_image/utils.py:15]: Not a big deal, but yeah, I think your GPU related functions are interesting enough to have their own module

**Comment by alexcombessie** [comment]
> > > ... and I found a UX 🐛 !
> > > Screen.Recording.2021-03-16.at.12.21.52.mov
> > > The GPU button cannot be deselected
> > > Source: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/recipes/compute_jb87eTC0_1_1/
> > 
> > 
> > About this, I can't replicate so I assume the refacto has fixed the issue by itself 💪
> 
> Hum let me test again before claiming the 🐻

Nope, the 🐻 is still alive, and says hi from a brand new recipe: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/recipes/compute_wfUkFdyl/

**Comment by HenriChabert** [comment]
> > > > ... and I found a UX 🐛 !
> > > > Screen.Recording.2021-03-16.at.12.21.52.mov
> > > > The GPU button cannot be deselected
> > > > Source: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/recipes/compute_jb87eTC0_1_1/
> > > 
> > > 
> > > About this, I can't replicate so I assume the refacto has fixed the issue by itself 💪
> > 
> > 
> > Hum let me test again before claiming the 🐻
> 
> Nope, the 🐻 is still alive, and says hi from a brand new recipe: http://dku11.dataiku.com:54321/projects/PLUGINTESTDEEPLEARNINGIMAGE/recipes/compute_wfUkFdyl/

Ohhh the smelly 🐻, I found it hiden in the woods. Fixed in 72a949e

**Review by alexcombessie — APPROVED** [review]
